documentation/src/main/asciidoc/user_guide/clustered_lock.adoc

```diff
@@ -0,0 +1,200 @@
+== Clustered Lock
+
+Clustered lock is a new building block introduced in Infinispan 9.2
+It is a lock distributed and shared among all nodes in the Infinispan cluster and currently provides a way to execute
+code that will be synchronized between the nodes in a given cluster.
+
+=== Installation
+
+In order to start using the clustered locks, you needs to add the dependency in your Maven `pom.xml` file:
+
+.pom.xml
+[source,xml]
+----
+<dependency>
+   <groupId>org.infinispan</groupId>
+   <artifactId>infinispan-clustered-lock</artifactId>
+   <version>...</version> <!-- 9.2.0.Final or higher -->
+</dependency>
+----
+
+Currently no extra configuration needs to be provided.
+
+=== The ClusteredLockConfiguration.
+
+Currently there is a single type of `ClusteredLock` supported : non reentrant, NODE ownership lock.
+
+==== Ownership
+
+* `NODE`
+When a `ClusteredLock` is defined, this lock can be used from all the nodes in the infinispan cluster.
+When the ownership is NODE type, this means that the owner of the lock is the infinispan node that acquired the lock
+at a given time.
+This means that each time we get a `ClusteredLock` instance with the `ClusteredCacheManager`, this instance will be the
+same instance for each infinispan node.
+This lock can be used to synchronize code between infinispan nodes. The advantage of this lock is that any thread in the
+node can release the lock at a given time.
+
+* `INSTANCE` - not yet supported
+
+When a `ClusteredLock` is defined, this lock can be used from all the nodes in the infinispan cluster.
+When the ownership is INSTANCE type, this means that the owner of the lock is the actual instance we acquired when
+`ClusteredLockManager.get("lockName")` is called.
+
+This means that each time we get a `ClusteredLock` instance with the `ClusteredCacheManager`, this instance will be
+a new instance.
```

Comment:
Sorry I'm commenting so late, but I find the change in the `ClusteredLockManager.get(name)` behavior confusing.

Personally I like the vert.x API, where acquiring the lock gives you a `Lock` instance, and you can only release it through that `Lock` instance. Being able to release a lock that some other application component has acquired without any communication with that component sounds wrong to me.

Reply:
Don't worry, input is always welcome Dan ! So as I understand, the ownership level "NODE" is not ok and we should support "INSTANCE" level locking (rentrant or not rentrant) ?

Reply:
About this point, @rvansa considered that we might only need "object" level (at leats for now). We discussed that briefly during the F2F

Reply:
First of all, I'm not enunciating rules here, I'm just thinking aloud :) But I think even with the vert.x API, node-ownership is still be useful if the lock is reentrant: that way multiple logical threads on the same node can acquire the lock concurrently, and only the final release actually allows another node to acquire it.

Thinking about it some more, maybe we should allow the application to use its own lock owner, as in `clusteredLock.lock(lockOwner)` or `clusteredLock.withOwner(lockOwner).lock()` (to match the not-yet-public API in Cache). That way, different application components that deal with a single request could call `get(name).lock(requestId)` without the need to explicitly pass a lock instance between them. Regular `lock()` would use the local node as the lock owner.

Reply:
@karesti It's you who came with the notion of node-level locks; I haven't seen that in Vert.X but the fact that the lock is not tied to a thread. To me a node-level ownership with non-reentrancy seems the same as instance-level. 

@danberindei Sounds good, but the owner should be provided when the lock is retrieved from manager rather than after that (to enforce const owner by api).

Reply:
@rvansa it's not accurate to say that was me who came to that idea of ownership or reentrancy levels, but this is not the discussion here.
I think there is a misunderstanding here and probably is due to something I did not fully get

Reply:
@danberindei concerning the "with owner" this is something I was considering too

Reply:
+1 for withOwner